app/workers/image_callback_worker.rb

```diff
@@ -43,7 +30,97 @@ def perform(_sqs_msg, job)
         announce(:series, :update, Api::Msg::SeriesRepresenter.new(series).to_json)
       end
     end
+  end
+
+  def perform_porter_job_callback(job)
+    job_id, task_type = job['JobResult']['Job']['Id'].split(':')
+    image = Image.by_porter_job_id(job_id)
+    if image.present?
+      case task_type
+      when 'copy'
```

Comment:
Ah, I see - so you're serializing the tasks on the rails worker side ... kicking the next one off when the previous succeeds.

My only concern there is it might be unpleasantly slow for Publish users, watching the progressbar as their image is "processing".  Any idea how long these tend to take?  (I think the worker only wakes up so often to check SQS).

The alternative would be a more binary `status` flow:

1. Fire all 3 tasks in a single porter job.  Set image.status = UPLOADED.
2. Listen for the first/any callback from porter.  Set image.status = VALIDATING (Porter received the job and is working on it).
3. Listen for the final overall job complete/error callback.  Check the task-results bundled in the callback json, and set the status to one of COMPLETE, FAILED (something went terribly wrong), INVALID (not an image), or NOTFOUND (404).

I can be talked into the serial job approach though.  It beats tracking 3 job_ids on each db record!

Reply:
My experience is that it's very fast, but that's running the worker locally which seems to pick jobs up much faster than that. Presumably the fact that the development queues aren't very busy has some impact there.

I don't really have a good way to avoid serializing Inspect => Thumbnail because I need the Format attribute from analysis to properly form the Porter jobs for generating thumbnails (unless I am misunderstanding something fundamental about Porter)

I guess I could assume based on filename, but that seems pretty fraught

I could run copy and inspect as tasks on the same Job, but unless I am misunderstanding tasks may be run in parallel and failures in a Map short circuit, which means that I couldn't be assured I would get back a failed copy task before I got a failed inspect task, which leaves some ambiguity between those two states (unless I do some string munging on the enclosed stack trace)

I would be very happy to be wrong about any part of this, so please let me know what you think!

Reply:
@farski please weigh in if I got part of this (or all of this) wrong!